### PR TITLE
[WIP] Multi stock 

### PR DESCRIPTION
**Pull request by umpirsky**

### CODE REVIEW

File: src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml [hunk 1]
```diff
@@ -74,7 +74,7 @@ fos_rest:
     format_listener:
         rules:
             - { path: '^/api', priorities: ['json', 'xml'], fallback_format: json, prefer_extension: true }
-            - { path: '^/', stop: true }
+            - { path: '^/' }
```
Comment (stloyd) [inline src/Sylius/Bundle/CoreBundle/Resources/config/app/main.yml]: Should be reverted.

File: src/Sylius/Bundle/CurrencyBundle/Form/Type/CurrencyCodeChoiceType.php [hunk 2]
```diff
@@ -42,7 +42,7 @@ public function setDefaultOptions(OptionsResolverInterface $resolver)
     {
         $choices = null;
 
-        foreach ($this->currencyProvider->getAvailableCurrencies() as $currency) {
+        foreach($this->currencyProvider->getAvailableCurrencies() as $currency) {
```
Comment (stloyd) [inline src/Sylius/Bundle/CurrencyBundle/Form/Type/CurrencyCodeChoiceType.php]: Should be reverted.

File: src/Sylius/Bundle/CurrencyBundle/spec/Templating/Helper/CurrencyHelperSpec.php [hunk 3]
```diff
@@ -21,11 +21,8 @@
  */
 class CurrencyHelperSpec extends ObjectBehavior
 {
-    function let(
-        CurrencyContextInterface $currencyContext,
-        CurrencyConverterInterface $converter,
-        MoneyHelper $moneyHelper
-    ) {
+    function let(CurrencyContextInterface $currencyContext, CurrencyConverterInterface $converter, MoneyHelper $moneyHelper)
+    {
```
Comment (stloyd) [inline src/Sylius/Bundle/CurrencyBundle/spec/Templating/Helper/CurrencyHelperSpec.php]: Should be reverted.

File: features/channels/channel_management.feature [hunk 4]
```diff
@@ -5,20 +5,22 @@ Feature: Channel management
     I want to configure channels
 
     Background:
-        Given store has default configuration
```
Comment (umpirsky) [inline features/channels/channel_management.feature]: Background needs fixing.

File: src/Sylius/Bundle/InventoryBundle/Doctrine/ORM/StockLocationRepository.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+class StockLocationRepository extends EntityRepository implements StockLocationRepositoryInterface
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function countBackorderableByStockable(StockableInterface $stockable)
+    {
+        return $this->createQueryBuilder('l')
+            ->leftJoin('l.stockItems', 'i')
+            ->select('COUNT(*)')
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/Doctrine/ORM/StockLocationRepository.php]: `COUNT(*)` is bad idea.

File: src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListenerInterface.php [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+namespace Sylius\Bundle\InventoryBundle\EventListener;
+
+use Symfony\Component\EventDispatcher\GenericEvent;
+
+/**
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+interface StockItemListenerInterface
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListenerInterface.php]: IMO this interface is useless.

File: src/Sylius/Bundle/InventoryBundle/Resources/config/driver/doctrine/mongodb-odm.xml [hunk 7]
```diff
@@ -1,56 +0,0 @@
-<?xml version="1.0" encoding="UTF-8"?>
```
Comment (stloyd) [inline src/Sylius/Bundle/InventoryBundle/Resources/config/driver/doctrine/mongodb-odm.xml]: Why?

File: src/Sylius/Bundle/ShippingBundle/Form/Type/Calculator/FlatRateConfigurationType.php [hunk 8]
```diff
@@ -52,7 +52,7 @@ public function buildForm(FormBuilderInterface $builder, array $options)
                 'constraints' => array(
                     new NotBlank(),
                     new Type(array('type' => 'integer')),
-                ),
+                )
```
Comment (stloyd) [inline src/Sylius/Bundle/ShippingBundle/Form/Type/Calculator/FlatRateConfigurationType.php]: Should be reverted.

File: src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/product.yml [hunk 9]
```diff
@@ -13,7 +13,6 @@ sylius_backend_product_index:
                 method: createFilterPaginator
                 arguments: [$criteria, $sorting, $deleted]
 
-# @todo refactor this when PRs about API & search get merged
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/config/routing/backend/product.yml]: This is still required IIRC.

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Payment/macros.html.twig [hunk 10]
```diff
@@ -41,7 +41,7 @@
                     {{ buttons.delete(path('sylius_backend_payment_delete', {'id': payment.id})) }}
                     <a href="{{ path('sylius_backend_payment_history', {'id': payment.id}) }}" class="btn btn-info">
                         <i class="glyphicon glyphicon-calendar"></i>
-                         {{ 'sylius.history'|trans }}
+                         {{ 'History'|trans }}
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Payment/macros.html.twig]: Should be reverted.

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Product/stock.html.twig [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+            <td class="center-text">
+                <br>
+                {{ variant }} ({{ variant.id }})
+                <hr>
+                <img src="{{ product.image ? product.image.path|imagine_filter('sylius_small') : 'http://placehold.it/50x40' }}" alt="" class="imgmedia-object" />
+            <td colspan="4">
+            <table class="table table-bordered">
+                <thead>
+                    <tr>
+                        <th>Stock Location</th>
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Product/stock.html.twig]: Should be translated. Same below.

File: src/Sylius/Bundle/WebBundle/Resources/views/Backend/Security/login.html.twig [hunk 12]
```diff
@@ -16,12 +16,13 @@
...
             {% stylesheets output='assets/compiled/backend_login.css' filter='cssrewrite'
-                '//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css'
                 '//netdna.bootstrapcdn.com/font-awesome/3.2.1/css/font-awesome.css'
-                'bundles/syliusweb/css/backend.css'
             %}
                 <link rel="stylesheet" href="{{ asset_url }}" type="text/css" />
             {% endstylesheets %}
+
+            <link rel="stylesheet" href="//netdna.bootstrapcdn.com/bootstrap/3.0.0/css/bootstrap.min.css" type="text/css" />
+            <link rel="stylesheet" href="{{ asset('bundles/syliusweb/css/backend.css') }}" type="text/css" />
```
Comment (stloyd) [inline src/Sylius/Bundle/WebBundle/Resources/views/Backend/Security/login.html.twig]: Why moving this here?

File: src/Sylius/Component/Core/Locale/LocaleContext.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
+<?php
```
Comment (stloyd) [inline src/Sylius/Component/Core/Locale/LocaleContext.php]: This file was removed in last refactoring.

File: src/Sylius/Component/Core/Model/ProductVariant.php [hunk 14]
```diff
@@ -173,9 +181,6 @@ public function getPrice()
      */
     public function setPrice($price)
     {
-        if (!is_int($price)) {
```
Comment (stloyd) [inline src/Sylius/Component/Core/Model/ProductVariant.php]: This looks wrong.

File: src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php [hunk 15]
```diff
@@ -91,6 +91,7 @@ public function processInventoryUnits(OrderItemInterface $item)
      */
     public function holdInventory(OrderInterface $order)
     {
+        return;
```
Comment (stloyd) [inline src/Sylius/Component/Core/OrderProcessing/InventoryHandler.php]: :stuck_out_tongue_winking_eye:

File: src/Sylius/Component/Inventory/Backorders/BackordersFillerInterface.php [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Inventory\Operator;
+
+use Doctrine\Common\Collections\Collection;
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Backorders/BackordersFillerInterface.php]: Never used.

File: src/Sylius/Component/Inventory/Backorders/BackordersManagerInterface.php [hunk 17]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Sylius\Component\Inventory\Operator;
+
+use Doctrine\Common\Collections\Collection;
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Backorders/BackordersManagerInterface.php]: Never used.

File: src/Sylius/Component/Inventory/Backorders/BackordersManagerInterface.php [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+ * Backorders manager interface.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+interface BackordersManagerInterface
+{
+    /**
+     * Is subject backorderable from any of available locations?
+     *
+     * @param Stockableinterface $stockable
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Backorders/BackordersManagerInterface.php]: Missing `@return bool`.

File: src/Sylius/Component/Inventory/Checker/AvailabilityChecker.php [hunk 19]
```diff
@@ -29,12 +35,12 @@ class AvailabilityChecker implements AvailabilityCheckerInterface
     protected $backorders;
 
     /**
-     * Constructor.
-     *
-     * @param bool $backorders
+     * @param bool                $backorders
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Checker/AvailabilityChecker.php]: This is second not first param.

File: src/Sylius/Component/Inventory/Factory/InventoryUnitFactoryInterface.php [hunk 20]
```diff
@@ -24,9 +24,9 @@
     /**
      * Create given amount of inventory units.
      *
-     * @param StockableInterface $stockable
-     * @param integer            $quantity
-     * @param string             $state
+     * @param StockableInterface     $stockable
+     * @param integer                $quantity
+     * @param string                 $state
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Factory/InventoryUnitFactoryInterface.php]: Should be reverted.

File: src/Sylius/Component/Inventory/Factory/StockItemFactory.php [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+    public function create(StockableInterface $stockable, StockLocationInterface $location)
+    {
+        if (null !== $item = $this->stockItemRepository->findByStockableAndLocation($stockable, $location)) {
+            return $item;
+        }
+
+        $item = $this->stockItemRepository->createNew();
+        $item
+            ->setStockable($stockable)
+            ->setLocation($location)
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Factory/StockItemFactory.php]: Don't use fluent interface.

File: src/Sylius/Component/Inventory/Factory/StockMovementFactory.php [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+    {
+        if (!is_integer($quantity) || 0 === $quantity) {
+            throw new \InvalidArgumentException('Invalid quantity given!');
+        }
+
+        $movement = $this->stockMovementRepository->createNew();
+
+        $movement
+            ->setStockItem($stockItem)
+            ->setQuantity($quantity)
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Factory/StockMovementFactory.php]: Don't use fluent interface.

File: src/Sylius/Component/Inventory/Model/StockItem.php [hunk 23]
```diff
@@ -0,0 +1,232 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setStockable(StockableInterface $stockable)
+    {
+        $this->stockable = $stockable;
+
+        return $this;
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Model/StockItem.php]: Please remove all fluent code.

File: src/Sylius/Component/Inventory/Model/StockLocation.php [hunk 24]
```diff
@@ -0,0 +1,223 @@
...
+    public function getName()
+    {
+        return $this->name;
+    }
+
+    public function setName($name)
+    {
+        $this->name = $name;
+
+        return $this;
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Model/StockLocation.php]: Same about fluent.

File: src/Sylius/Component/Inventory/Model/StockMovement.php [hunk 25]
```diff
@@ -0,0 +1,116 @@
...
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function setStockItem(StockItemInterface $stockItem)
+    {
+        $this->stockItem = $stockItem;
+
+        return $this;
```
Comment (stloyd) [inline src/Sylius/Component/Inventory/Model/StockMovement.php]: Again fluent.

File: src/Sylius/Component/Payment/Model/PaymentInterface.php [hunk 26]
```diff
@@ -25,7 +25,7 @@
     const STATE_NEW        = 'new';
     const STATE_PENDING    = 'pending';
     const STATE_PROCESSING = 'processing';
-    const STATE_COMPLETED  = 'completed';
+    const STATE_COMPLETED  = 'complete';
```
Comment (stloyd) [inline src/Sylius/Component/Payment/Model/PaymentInterface.php]: Should be reverted?

**Comment by kayue** [comment]
@pjedrzejewski What do you feel about this? Would you merge this as a whole (if everything is fixed), or should we break this into smaller PR?

**Comment by pjedrzejewski** [comment]
@kayue I think breaking it into smaller PRs would be good idea, cause as we see big PRs are root of all evil and delay. :) On the other hand, not sure what we could split here, it is pretty much connected with everything... Ideas?

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 27]
```diff
@@ -152,6 +166,33 @@ public function findForDetailsPage($id)
     }
 
     /**
+     * Get the product data for the stock page.
+     *
+     * @param integer $id
+     *
+     * @return null|ProductInterface
+     */
+    public function findForStockPage($id)
```
Comment (umpirsky) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: Repository should now know about pages.

File: src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php [hunk 28]
```diff
@@ -152,6 +166,33 @@ public function findForDetailsPage($id)
...
+    {
+        $this->_em->getFilters()->disable('softdeleteable');
+
+        $queryBuilder = $this->getQueryBuilder();
+        $queryBuilder
+            ->leftJoin('variant.stockItems', 'stockItem')
+            ->addSelect('stockItem')
+            ->leftJoin('stockItem.location', 'stockLocation')
+            ->addSelect('stockLocation')
+            ->andWhere($queryBuilder->expr()->eq('product.id', ':id'))
```
Comment (umpirsky) [inline src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php]: Is there any advantage of using expr over `=`?

File: src/Sylius/Bundle/CurrencyBundle/Form/Type/CurrencyEntityChoiceType.php [hunk 29]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+namespace Sylius\Bundle\CurrencyBundle\Form\Type;
+
+/**
+ * Currency entity choice form.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class CurrencyEntityChoiceType extends CurrencyChoiceType
```
Comment (umpirsky) [inline src/Sylius/Bundle/CurrencyBundle/Form/Type/CurrencyEntityChoiceType.php]: Can be removed.

File: src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListener.php [hunk 30]
```diff
@@ -0,0 +1,52 @@
...
+     */
+    public function __construct(StockItemFactoryInterface $stockItemFactory)
+    {
+        $this->stockItemFactory = $stockItemFactory;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function onStockableCreate(GenericEvent $event)
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/EventListener/StockItemListener.php]: Name based on functionality, not event.

File: src/Sylius/Bundle/InventoryBundle/Form/Type/StockItemType.php [hunk 31]
```diff
@@ -0,0 +1,45 @@
...
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $builder
+            ->add('availableOnDemand', 'checkbox', array(
+                'required' => false,
+                'label'    => 'sylius.form.stock_item.available_on_demand',
+            ))
+        ;
+
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/Form/Type/StockItemType.php]: Extra line.

File: src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationEntityType.php [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+namespace Sylius\Bundle\InventoryBundle\Form\Type;
+
+/**
+ * StockLocation entity choice form.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class StockLocationEntityType extends StockLocationChoiceType
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationEntityType.php]: Can be removed.

File: src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationType.php [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+            ))
+            ->add('name', 'text', array(
+                'label' => 'sylius.form.stock_location.name'
+            ))
+            ->add('enabled', 'checkbox', array(
+                'required' => false,
+                'label'    => 'sylius.form.stock_location.enabled',
+            ))
+        ;
+
```
Comment (umpirsky) [inline src/Sylius/Bundle/InventoryBundle/Form/Type/StockLocationType.php]: Extra line.

File: src/Sylius/Bundle/LocaleBundle/Form/Type/LocaleEntityChoiceType.php [hunk 34]
```diff
@@ -0,0 +1,28 @@
...
+ */
+
+namespace Sylius\Bundle\LocaleBundle\Form\Type;
+
+/**
+ * Locale entity choice form.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class LocaleEntityChoiceType extends LocaleChoiceType
```
Comment (umpirsky) [inline src/Sylius/Bundle/LocaleBundle/Form/Type/LocaleEntityChoiceType.php]: Can be removed.

File: src/Sylius/Component/Channel/Model/ChannelType.php [hunk 35]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+namespace Sylius\Component\Channel\Model;
+
+/**
+ * Channel types.
+ *
+ * @author Paweł Jędrzejewski <pawel@sylius.org>
+ */
+class ChannelType
```
Comment (umpirsky) [inline src/Sylius/Component/Channel/Model/ChannelType.php]: Can be removed.

File: src/Sylius/Component/Inventory/Model/StockItem.php (hunk 23 above)
Comment (umpirsky) [inline src/Sylius/Component/Inventory/Model/StockItem.php]: Why?

Comment (stloyd) [inline src/Sylius/Component/Inventory/Model/StockItem.php]: As this it's done in everywhere other touched places, cause ["fluent interfaces are evil"](http://ocramius.github.io/blog/fluent-interfaces-are-evil/) :wink:

**Comment by umpirsky** [comment]
@pjedrzejewski @michalmarcinkowski @kayue This is pretty fat PR, and we all know PRs like this become outdated pretty easy and never reach master. We don't want to do that, we want to contribute something that is atomic and useful.

In order to make it easier to track, I can recognize several major changegroups in this PR:

- Password resetting.
- Stock locations.
- Stock items.
- Stock movement.
- Stock transfer.
- Quantifier.
- Coordinator
- Packer.
- Splitter.
- Backorder filler.
- Location based pricing calculator.
- Inventory unit handling changes.
- Inventory operator, availability checker changes.
- Taxon API.
- Checkout purchase transition.
- Payment changes.

The first step we would like to extract several of them and group them into one less fat (still big) PR:

- Stock locations.
- Stock items.
- Stock movement.
- Stock transfer.
- Quantifier.
- Coordinator
- Packer.
- Splitter.
- Backorder filler.
- Inventory operator, availability checker changes.

Do you think you will be able to merge something like this once it's completed? What would be your requirements in order to make it mater ready?
I think basic Behat scenarios to make sure we are all on the same page and clean code. I would also appreciate community feedback, and I am really happy we already have some (big thanks @stloyd).

**Comment by kayue** [comment]
@pjedrzejewski Basically at the first step we want to focus on bringing the `StockItem` model to core. This is an important architecture change, the sooner we have this in core the better for everyone. 

New features like "Location based pricing calculator" can be added later on. We want to focus on updating the architecture without breaking things at the first step. We hope this will make the review easier.

Kindly let us know is this something you would merge and include it in the next version?

File: src/Sylius/Bundle/CoreBundle/EventListener/OrderPricingListener.php [hunk 36]
```diff
@@ -63,6 +63,10 @@ public function recalculatePrices(GenericEvent $event)
             $context['channel'] = array($order->getChannel());
         }
 
+        if (null !== $order->getChannel()) {
```
Comment (umpirsky) [inline src/Sylius/Bundle/CoreBundle/EventListener/OrderPricingListener.php]: Duplicated.

**Comment by kayue** [comment]
ping @pjedrzejewski for feedback :)

**Comment by pjedrzejewski** [comment]
Hey @kayue and @umpirsky!

I'd drop "password resetting" from that list. It was quick implementation to expose this as API, let's focus on multi-stock here.

I think it makes sense to do it in the steps outlined above. :) First 4 points could be done in smaller steps, but rest strongly affects the inventory handling, so I think it should be done all together.

I'd skip any stuff that is not required for initial version to work, so this pricing calculator and any extra stuff should be left for later.

Let's look at first step - StockItem and StockLocation changes:

* How we will decide which stock item should be related to inventory unit? Or actually, maybe for this first iteration, we still keep InventoryUnit -> ProductVariant (stockable) relation and do that integration later. Maybe for first step we just introduce stock item and stock location management in the backend.

Then go to stock movement (updating quantities in the backend) and finally drop "onHand" and "onHold" fields from variants and integrate them with stock items? I believe this is what you had in mind, right?

Looking forward to your feedbakc. :)

**Comment by kayue** [comment]
@pjedrzejewski Sounds good. 

We will start with introduce stock item and stock location at the first step, and won't touch how quantity is handle at this stage. 

Once our first PR is merged we will start stock movement and quantity handling shortly.

Thanks again!

**Comment by pjedrzejewski** [comment]
@kayue @umpirsky :+1:

**Pull request closed without merging**